### ISSUE TITLE
Broken support for FQDN with trailing dot '.'

### ISSUE THREAD

**Issue by lmarszal** [istio/istio#28103]
**Bug description**
Istio 1.5 introduced support for FQDN with trailing dot '.' (https://github.com/istio/istio/issues/13710)
This feature is broken in istio 1.6 and istio 1.7 because of this bug: https://github.com/envoyproxy/envoy/issues/13587. Also produced cluster entries and route configs looks odd.

**Steps to reproduce the bug**  

**First test:**
`Service` obj with `type: ExternalName`:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: wikipedia
spec:
  type: ExternalName
  externalName: wikipedia.org.
  ports:
  - name: http
    port: 80
    protocol: TCP
```
Envoy can't resolve this DNS name because of https://github.com/envoyproxy/envoy/issues/13587 bug, but cluster and route config looks OK.  
  
Cluster:
```json
    {
     "version_info": "2020-10-20T11:48:12Z/3",
     "cluster": {
      "@type": "type.googleapis.com/envoy.config.cluster.v3.Cluster",
      "name": "outbound|80||wikipedia.default.svc.cluster.local",
      "type": "STRICT_DNS",
      "connect_timeout": "10s",
      "circuit_breakers": {
       "thresholds": [
        {
         "max_connections": 4294967295,
         "max_pending_requests": 4294967295,
         "max_requests": 4294967295,
         "max_retries": 4294967295
        }
       ]
      },
      "dns_refresh_rate": "5s",
      "dns_lookup_family": "V4_ONLY",
      "load_assignment": {
       "cluster_name": "outbound|80||wikipedia.default.svc.cluster.local",
       "endpoints": [
        {
         "locality": {},
         "lb_endpoints": [
          {
           "endpoint": {
            "address": {
             "socket_address": {
              "address": "wikipedia.org.",
              "port_value": 80
             }
            }
           },
           "metadata": {},
           "load_balancing_weight": 1
          }
         ],
         "load_balancing_weight": 1
        }
       ]
      },
      "respect_dns_ttl": true,
      "filters": [
       {
        "name": "istio.metadata_exchange",
        "typed_config": {
         "@type": "type.googleapis.com/udpa.type.v1.TypedStruct",
         "type_url": "type.googleapis.com/envoy.tcp.metadataexchange.config.MetadataExchange",
         "value": {
          "protocol": "istio-peer-exchange"
         }
        }
       }
      ]
     },
     "last_updated": "2020-10-20T11:48:12.029Z"
    }
```
Route config:
```json
       {
        "name": "wikipedia.default.svc.cluster.local:80",
        "domains": [
         "wikipedia.default.svc.cluster.local",
         "wikipedia.default.svc.cluster.local:80",
         "wikipedia",
         "wikipedia:80",
         "wikipedia.default.svc.cluster",
         "wikipedia.default.svc.cluster:80",
         "wikipedia.default.svc",
         "wikipedia.default.svc:80",
         "wikipedia.default",
         "wikipedia.default:80"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/"
          },
          "route": {
           "cluster": "outbound|80||wikipedia.default.svc.cluster.local",
           "timeout": "0s",
           "retry_policy": {
            "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
            "num_retries": 2,
            "retry_host_predicate": [
             {
              "name": "envoy.retry_host_predicates.previous_hosts"
             }
            ],
            "host_selection_retry_max_attempts": "5",
            "retriable_status_codes": [
             503
            ]
           },
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "wikipedia.default.svc.cluster.local:80/*"
          },
          "name": "default"
         }
        ],
        "include_request_attempt_count": true
       }
```

**Second test:**
`ServiceEntry` with `resolution: DNS`:
```yaml
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: wikipedia
spec:
  hosts:
  - wikipedia.org.
  location: MESH_EXTERNAL
  resolution: DNS
  ports:
  - name: http
    number: 80
    protocol: HTTP
```
Still not working in envoy, but cluster name and route domains looks odd too.  
  
Cluster:
```json
    {
     "version_info": "2020-10-20T12:00:11Z/5",
     "cluster": {
      "@type": "type.googleapis.com/envoy.config.cluster.v3.Cluster",
      "name": "outbound|80||wikipedia.org.",
      "type": "STRICT_DNS",
      "connect_timeout": "10s",
      "circuit_breakers": {
       "thresholds": [
        {
         "max_connections": 4294967295,
         "max_pending_requests": 4294967295,
         "max_requests": 4294967295,
         "max_retries": 4294967295
        }
       ]
      },
      "dns_refresh_rate": "5s",
      "dns_lookup_family": "V4_ONLY",
      "load_assignment": {
       "cluster_name": "outbound|80||wikipedia.org.",
       "endpoints": [
        {
         "locality": {},
         "lb_endpoints": [
          {
           "endpoint": {
            "address": {
             "socket_address": {
              "address": "wikipedia.org.",
              "port_value": 80
             }
            }
           },
           "load_balancing_weight": 1
          }
         ],
         "load_balancing_weight": 1
        }
       ]
      },
      "respect_dns_ttl": true,
      "filters": [
       {
        "name": "istio.metadata_exchange",
        "typed_config": {
         "@type": "type.googleapis.com/udpa.type.v1.TypedStruct",
         "type_url": "type.googleapis.com/envoy.tcp.metadataexchange.config.MetadataExchange",
         "value": {
          "protocol": "istio-peer-exchange"
         }
        }
       }
      ]
     },
     "last_updated": "2020-10-20T12:00:11.031Z"
    }
```
Route config:
```json
       {
        "name": "wikipedia.org.:80",
        "domains": [
         "wikipedia.org.",
         "wikipedia.org.:80"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/"
          },
          "route": {
           "cluster": "outbound|80||wikipedia.org.",
           "timeout": "0s",
           "retry_policy": {
            "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
            "num_retries": 2,
            "retry_host_predicate": [
             {
              "name": "envoy.retry_host_predicates.previous_hosts"
             }
            ],
            "host_selection_retry_max_attempts": "5",
            "retriable_status_codes": [
             503
            ]
           },
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "wikipedia.org.:80/*"
          },
          "name": "default"
         }
        ],
        "include_request_attempt_count": true
       }
```

**Expected behavior**
In cluster generated for `ServiceEntry` cluster name should be `outbound|80||wikipedia.org` (no dot).  
Cluster's `lb_endpoint` address should remain `wikipedia.org.` (believe this should be fixed upstream, in envoy).
In route config generated for `ServiceEntry` route name should be `wikipedia.org:80` (no dot).    
In route config generated for `ServiceEntry` domains should include both trailing dot and not:
```json
[
  "wikipedia.org.",
  "wikipedia.org.:80",
  "wikipedia.org",
  "wikipedia.org:80"
]
```

This might also impact `DestinationRule` and `VirtualService` generation - that's why I want generated names not to have dot at the end.
  
  
  
Since this is not clearly documented, I don't know if this still counts as a bug or as new feature. But - with upstream fix in envoy - this seems to be low hanging fruit for limiting redundant DNS queries when having lots of strict_dns clusters in your mesh.

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Version**
```
$ istioctl version --remote
client version: 1.7.3
control plane version: 1.7.3
data plane version: 1.7.3 (1 proxies)
$ kubectl version --short
Client Version: v1.18.8
Server Version: v1.17.9
```

**How was Istio installed?**
```
$ istioctl install --set profile=minimal
$ kubectl label namespace default istio-injection=enabled
```

**Environment where bug was observed (cloud vendor, OS, etc)**
```
$ minikube version
minikube version: v1.13.1
commit: 1fd1f67f338cbab4b3e5a6e4c71c551f522ca138
```

**Comment by howardjohn** [comment]
Hey, could you expand more a bit on why you think
> In route config generated for ServiceEntry domains should include both trailing dot and not:

I did some research on this in the past and did not find any relevant standards suggesting this behavior

**Comment by lmarszal** [comment]
> Hey, could you expand more a bit on why you think
> 
> > In route config generated for ServiceEntry domains should include both trailing dot and not:
> 
> I did some research on this in the past and did not find any relevant standards suggesting this behavior

This is just because of this comment of yours https://github.com/istio/istio/issues/13710#issuecomment-494477707. If this is not valid anymore, then route config probably should include only domain names without a dot.

**Comment by howardjohn** [comment]
Right, my point there was it seems like every client handles the dot different, and on the other side every server *also* handles it different, so its not clear what the "correct" behavior is here.

**Comment by lmarszal** [comment]
You may be right. Let me elaborate more to see if I understand you:

**Making TCP connection (istio-proxy->server):**
TCP uses only IP address - so it doesn't matter how you address your endpoint, as long as it correctly resolves.
It looks like established pattern, that DNS resolvers in unix/linux accepts trailing dot syntax as unambiguous version of domain name.

**Making TLS handshake over that connection (if appplicable):**
Looks like SNI expects hostname without trailing dot ([RFC6066](https://tools.ietf.org/html/rfc6066#section-3)):
> The hostname is represented as a byte string using ASCII encoding without a trailing dot.

Although specific client/server implementation details are vague. Luckily you [can specify sni in DestinationRule](https://istio.io/latest/docs/reference/config/networking/destination-rule/#ClientTLSSettings-sni).

**HTTP Host header (or equivalent in other protocols):**
I would guess most client/server libraries sends hostname without a trailing dot - but again (as you said) this might be different between implementations.  
**(istio-proxy->server):** Host headers are set independently in [HttpRewrite section of VirtualService](https://istio.io/latest/docs/reference/config/networking/virtual-service/#HTTPRewrite).  
**(app->istio-proxy):** I think there's no other option but accept that some client will send hostnames with dot at the end. The question here is (1) if `wikipedia.org.` is equal `wikipedia.org` or they should be routed independently (2) should this "accept both" behaviour be opt-in or default.


What I'm really fighting for is using name with trailing dot for endpoint dns name resolution. This can be achieved also by `endpoints` field of ServiceEntry:
```yaml
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: wikipedia
spec:
  hosts:
  - wikipedia.org
  endpoints:
  - address: wikipedia.org.
  location: MESH_EXTERNAL
  ports:
  - name: http
    number: 80
    protocol: HTTP
  resolution: DNS
```
But this (existing) syntax might be better than building logic around trailing dot in host field.

**Comment by lmarszal** [comment]
Just for the record - upstream change, fixing envoy support for `STRICT_DNS` cluster having endpoint address with trailing dot just got merged to master: https://github.com/envoyproxy/envoy/pull/13644

**Comment by howardjohn** [comment]
Not stale

**Comment by aaronmell** [comment]
Does this work now in 1.9?

**Comment by derekcrovo** [comment]
We hit this old issue here in October 2021. If you're seeing it too and don't like the idea of doubling all your VirtualServices and ServiceEntries, try this EnvoyFilter. It toggles the Envoy `strip_trailing_host_dot` feature. It completely solved the problem for us.


```yaml
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: strip-trailing-dot
  namespace: istio-system
spec:
  configPatches:
  - applyTo: NETWORK_FILTER
    match:
      listener:
        filterChain:
          filter:
            name: "envoy.filters.network.http_connection_manager"
    patch:
      operation: MERGE
      value:
        typed_config:
          "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager"
          strip_trailing_host_dot: true
```

🎃

**Comment by kevanslumin** [comment]
@derekcrovo what Istio version did you hit the issue on?  Also, what did you mean by having to duplicate VirtualServices and ServiceEntries? It looks like Istio 1.10 has Envoy 1.18 which should include the fix.

**Comment by derekcrovo** [comment]
We're running 1.11.3 at the moment. I didn't see a way to enable this setting with Istio config, but the good ol' break glass API with EnvoyFilters worked great. I think the fix landed in Envoy 1.19, and 1.11 uses Envoy 1.19.2.

The previous workaround people had suggested is to have 2 VirtualServices/ServiceEntries for each host; one with the regular name and one with the name followed by the trailing dot. This will work, of course, since it will add those entries to Istio's catalog, but it doubles the size of the catalog.
